File: frontend/types/bacs.ts
export type BacsDomain = "monitoring" | "heating" | "cooling_ventilation" | "dhw" | "lighting";
export type BacsClass = "A" | "B" | "C" | "D" | "E";

export type BacsFunctionResponse = {
  id: string;
  code: string;
  domain: BacsDomain;
  name: string;
  description: string | null;
  weight: number;
  order_index: number;
  is_selected: boolean;
};

export type BacsCurrentResponse = {
  assessment_id: string | null;
  project_id: string;
  version: string;
  assessor_name: string | null;
  manual_override_class: BacsClass | null;
  notes: string | null;
  functions: BacsFunctionResponse[];
};

export type BacsDomainScoreResponse = {
  domain: BacsDomain;
  score: number;
  selected_weight: number;
  total_weight: number;
};

export type BacsMissingFunctionResponse = {
  id: string;
  code: string;
  domain: BacsDomain;
  name: string;
  description: string | null;
  weight: number;
};

export type BacsSummaryResponse = {
  assessment_id: string | null;
  project_id: string;
  version: string;
  confidence_score: number;
  overall_score: number;
  estimated_bacs_class: BacsClass;
  manual_override_class: BacsClass | null;
  bacs_class: BacsClass;
  selected_function_count: number;
  total_function_count: number;
  domain_scores: BacsDomainScoreResponse[];
  top_missing_functions: BacsMissingFunctionResponse[];
};

export type BacsAssessmentUpsertPayload = {
  assessor_name: string | null;
  manual_override_class: BacsClass | null;
  notes: string | null;
};

export type BacsCurrentFunctionsUpdatePayload = {
  selected_function_ids: string[];
};

File: backend/app/schemas/bacs.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

BacsDomain = Literal["monitoring", "heating", "cooling_ventilation", "dhw", "lighting"]
BacsClass = Literal["A", "B", "C", "D", "E"]


class BacsAssessmentUpsert(BaseModel):
    assessor_name: str | None = Field(default=None, max_length=255)
    manual_override_class: BacsClass | None = None
    notes: str | None = None


class BacsCurrentFunctionsUpdate(BaseModel):
    selected_function_ids: list[UUID] = Field(default_factory=list)


class BacsFunctionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    domain: BacsDomain
    name: str
    description: str | None
    weight: float
    order_index: int
    is_selected: bool


class BacsCurrentResponse(BaseModel):
    assessment_id: UUID | None
    project_id: UUID
    version: str
    assessor_name: str | None
    manual_override_class: BacsClass | None
    notes: str | None
    functions: list[BacsFunctionResponse]


class BacsDomainScoreResponse(BaseModel):
    domain: BacsDomain
    score: float
    selected_weight: float
    total_weight: float


class BacsMissingFunctionResponse(BaseModel):
    id: UUID
    code: str
    domain: BacsDomain
    name: str
    description: str | None
    weight: float


class BacsSummaryResponse(BaseModel):
    assessment_id: UUID | None
    project_id: UUID
    version: str
    confidence_score: float
    overall_score: float
    estimated_bacs_class: BacsClass
    manual_override_class: BacsClass | None
    bacs_class: BacsClass
    selected_function_count: int
    total_function_count: int
    domain_scores: list[BacsDomainScoreResponse]
    top_missing_functions: list[BacsMissingFunctionResponse]
